### PR TITLE
Add unique id for spot task and --no-follow option for logs

### LINKED ISSUE #1095: [Spot] Unique run id for different attempts after preemption
We should provide a unique run_id for the same spot job across different attempts so that the user can use that id to communicate with wandb and any other software that needs the same identification.

### PR DESCRIPTION
**Pull request by Michaelvll**
Closes #1095

Tested:
- [x] `sky spot launch -n test-bqa --cloud gcp --env WANDB_API_KEY ./examples/spot/bert_qa.yaml`
- [ ] `./tests/run_smoke_tests.sh test_spot_recovery`

### CODE REVIEW

File: sky/cli.py [hunk 1]
```diff
@@ -1328,14 +1328,24 @@ def queue(clusters: Tuple[str], skip_finished: bool, all_users: bool):
     default=False,
     help=('If specified, do not show logs but exit with a status code for the '
           'job\'s status: 0 for succeeded, or 1 for all other statuses.'))
+@click.option('--follow/--no-follow',
+              is_flag=True,
+              default=True,
+              help='Follow the logs of the job. [default: --follow]')
```
Comment (concretevitamin) [inline sky/cli.py:1334]: Maybe add: `--no-follow to print the log so far and exit.`.

File: sky/skylet/log_lib.py [hunk 2]
```diff
@@ -403,7 +404,9 @@ def tail_logs(job_owner: str,
         time.sleep(_SKY_LOG_WAITING_GAP_SECONDS)
         status = job_lib.update_job_status(job_owner, [job_id], silent=True)[0]
 
-    if status in [job_lib.JobStatus.RUNNING, job_lib.JobStatus.PENDING]:
+    if follow and status in [
```
Comment (concretevitamin) [inline sky/skylet/log_lib.py:407]: Maybe worth to document this arg in this func's docstr. Not entirely sure my other comment is correct.

File: docs/source/examples/spot-jobs.rst [hunk 3]
```diff
@@ -147,7 +148,8 @@ With the above changes, you are ready to launch a spot job with ``sky spot launc
     $ sky spot launch -n bert-qa bert_qa.yaml
 
 SkyPilot will launch and start monitoring the spot job. When a preemption happens, SkyPilot will automatically
-search for resources across regions and clouds to re-launch the job.
+search for resources across regions and clouds to re-launch the job. The environment variable :code:`$SKYPILOT_SPOT_RUN_ID`
+can be used to identify the same run as it is identical across multiple preemptions/re-launchs for the same spot job.
```
Comment (concretevitamin) [inline docs/source/examples/spot-jobs.rst:152]: nit:

> To identify the same job across multiple recoveries, use the environment variable ``$SKYPILOT_SPOT_RUN_ID`` in the task's `run` commands or directly in the program itself (e.g., access via ``os.environ`` and pass to Weights & Biases for tracking purposes in your training script). It is made available to the task whenever it is invoked. The spot run ID is kept identical  across all recoveries of the same spot job.

Comment (concretevitamin) [reply docs/source/examples/spot-jobs.rst:152]: A question as I wrote the above -- what's the reason to write the glue code e.g., L131, rather than directly using ``os.environ`` in the train script? Is it because the former is more visible?

File: sky/spot/spot_state.py [hunk 4]
```diff
@@ -99,7 +101,7 @@ def set_submitted(job_id: int, name: str, run_timestamp: str,
         status=(?),
         run_timestamp=(?)
         WHERE job_id=(?)""",
-        (name, resources_str, time.time(), SpotStatus.SUBMITTED.value,
+        (name, resources_str, submit_time, SpotStatus.SUBMITTED.value,
```
Comment (concretevitamin) [inline sky/spot/spot_state.py:104]: Hmm, what's the reason to use this earlier time vs. `time.time()`?  

This rings a bell because the table schema has `submitted_at` and `run_timestamp`. Now we're making both columns store the same time.

File: sky/spot/controller.py [hunk 5]
```diff
@@ -29,10 +29,18 @@ def __init__(self, job_id: int, task_yaml: str,
...
+
         self._retry_until_up = retry_until_up
         # TODO(zhwu): this assumes the specific backend.
         self.backend = cloud_vm_ray_backend.CloudVmRayBackend()
 
+        # Add a unique identifier to the task environment variables, so that
+        # the user can have the same id for multiple recoveries.
+        task_envs = self._task.envs or {}
+        task_envs['SKYPILOT_SPOT_RUN_ID'] = (f'{self.backend.run_timestamp}'
+                                             f'_id-{self._job_id}')
```
Comment (concretevitamin) [inline sky/spot/controller.py:41]: nit: show an example value? Maybe change `_id-` -> `_` to make it shorter?

**Review by concretevitamin — COMMENTED** [review]
Nice @Michaelvll - a few questions.

File: sky/spot/spot_state.py (hunk 4 above)
Comment (Michaelvll) [reply sky/spot/spot_state.py:104]: The variable name `run_timestamp` is a bit misleading, whose actual meaning is the run_id for the run `sky-2022-10-04....`, i.e. the logs will be stored in f`~/sky_logs/{run_timestamp}`.

Here I want to make the timestamp for submission, sky_logs and job id align with each other, so as to make it easier to find them based on one of the values. 

Also, using the earlier timestamp should be closer to the term `submit_at`, which should represent the time the spot task is submitted.

File: sky/spot/controller.py (hunk 5 above)
Comment (Michaelvll) [reply sky/spot/controller.py:41]: An example value would be: `sky-2022-10-04-22-46-52-467694_id-17`. I feel like withhout the `_id`, it will be a bit confusing what the last number is at the end. Wdyt?

File: docs/source/examples/spot-jobs.rst (hunk 3 above)
Comment (Michaelvll) [reply docs/source/examples/spot-jobs.rst:152]: The `--run_name` is natively supported by the huggingface repo, so there is no need to modify the code by using that option.
The run_name can definitely written in the code directly by using `os.environ`. I think one advantage for passing it through the argument is that, when the user wants to run it on on-demand VM, she can easily modify this `run_name` to any other costant name in the YAML file without changing it in the training code.

File: sky/cli.py (hunk 1 above)
Comment (Michaelvll) [reply sky/cli.py]: Added. Thanks!

File: docs/source/examples/spot-jobs.rst (hunk 3 above)
Comment (Michaelvll) [reply docs/source/examples/spot-jobs.rst]: Fixed the doc. Thanks!

File: sky/skylet/log_lib.py (hunk 2 above)
Comment (Michaelvll) [reply sky/skylet/log_lib.py:415]: Added. Thanks!

File: docs/source/examples/spot-jobs.rst (hunk 3 above)
Comment (concretevitamin) [reply docs/source/examples/spot-jobs.rst]: > I think one advantage for passing it through the argument is that, when the user wants to run it on on-demand VM, she can easily modify this run_name to any other costant name in the YAML file without changing it in the training code.

This reminds of the following: maybe we should make the run ID available regardless of on-demand or spot jobs. This way, no need for users to change their yaml? Can defer to a TODO.

**Review by concretevitamin — APPROVED** [review]

File: docs/source/examples/spot-jobs.rst (hunk 3 above)
Comment (Michaelvll) [reply docs/source/examples/spot-jobs.rst]: Good point! I renamed the variable to `SKYPILOT_RUN_ID` so that the on-demand can use the same name in the future.
Added a TODO in the controller.py

File: sky/spot/spot_state.py (hunk 4 above)
Comment (Michaelvll) [reply sky/spot/spot_state.py:109]: Added the comment. Thanks!

File: sky/spot/controller.py (hunk 5 above)
Comment (Michaelvll) [reply sky/spot/controller.py]: Done. Thanks!

File: sky/cli.py [hunk 6]
```diff
@@ -1328,10 +1328,12 @@ def queue(clusters: Tuple[str], skip_finished: bool, all_users: bool):
...
-@click.option('--follow/--no-follow',
-              is_flag=True,
-              default=True,
-              help='Follow the logs of the job. [default: --follow]')
+@click.option(
+    '--follow/--no-follow',
+    is_flag=True,
+    default=True,
+    help=('Follow the logs of the job. [default: --follow]'
+          'If --no-follow is specified, print the log so far and exit.'))
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/cli.py:1336]: Done. Thanks!

File: docs/source/examples/spot-jobs.rst [hunk 7]
```diff
@@ -148,8 +148,11 @@ With the above changes, you are ready to launch a spot job with ``sky spot launc
     $ sky spot launch -n bert-qa bert_qa.yaml
 
 SkyPilot will launch and start monitoring the spot job. When a preemption happens, SkyPilot will automatically
-search for resources across regions and clouds to re-launch the job. The environment variable :code:`$SKYPILOT_SPOT_RUN_ID`
-can be used to identify the same run as it is identical across multiple preemptions/re-launchs for the same spot job.
+search for resources across regions and clouds to re-launch the job. To identify the same job across multiple
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply docs/source/examples/spot-jobs.rst]: Good point! Added the "To identify...` in a NOTE block, which looks a bit nicer to me than the first option.

**Comment by Michaelvll** [comment]
Tested:
- [x] `sky spot launch -n test-bqa --cloud gcp --env WANDB_API_KEY ./examples/spot/bert_qa.yaml`
- [x] `./tests/run_smoke_tests.sh test_spot_recovery`
